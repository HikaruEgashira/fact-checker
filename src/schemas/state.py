import os
from typing import Literal, Union
import boto3
from aws_lambda_powertools.utilities.parser import BaseModel


TABLE_NAME = os.environ.get("TABLE_NAME") or "fact-checker-results"

dynamodb = boto3.resource("dynamodb")
table = dynamodb.Table(TABLE_NAME)

FactCheckOutput = str
Status = Union[Literal["pending"], Literal["completed"]]

_session_id: str


def current_session():
    return _session_id


class State(BaseModel):
    id: str
    status: Status
    output: FactCheckOutput


def update_state(state: State):
    global _session_id
    _session_id = state.id
    item = state.model_dump()
    return table.put_item(Item=item)


def get_state(state_id: str):
    global _session_id
    _session_id = state_id
    response = table.get_item(Key={"id": state_id})
    return State(**response["Item"]) if "Item" in response else None  # type: ignore


# only for testing purposes
def delete_state(state_id: str):
    return table.delete_item(Key={"id": state_id})
