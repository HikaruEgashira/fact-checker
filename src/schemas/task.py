import os
from typing import Literal, Union
import boto3
from aws_lambda_powertools.utilities.parser import BaseModel


TABLE_NAME = os.environ.get("TABLE_NAME") or "fact-checker-results"

dynamodb = boto3.resource("dynamodb")
table = dynamodb.Table(TABLE_NAME)

FactCheckOutput = Literal["accurate", "inaccurate", "false", "indeterminate"]
Status = Union[Literal["pending"], FactCheckOutput]


class Task(BaseModel):
    task_id: str
    text: str
    result: Status


def update_task(task: Task):
    return table.put_item(Item=task.model_dump())


def get_task(task_id: str):
    response = table.get_item(Key={"task_id": task_id})
    return Task(**response["Item"]) if "Item" in response else None  # type: ignore


# only for testing purposes
def delete_task(task_id: str):
    return table.delete_item(Key={"task_id": task_id})
