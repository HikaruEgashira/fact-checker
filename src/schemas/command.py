import os
from typing import Literal, Union
import boto3
from aws_lambda_powertools.utilities.parser import BaseModel, Field

from schemas.state import current_session

QUEUE_NAME = os.environ.get("QUEUE_NAME") or "fact-checker-queue"
REGION = os.environ.get("AWS_REGION") or "ap-northeast-1"

session = boto3.Session()
sqs = session.client("sqs")
sts = session.client("sts")

queue_url = f"https://sqs.{REGION}.amazonaws.com/{sts.get_caller_identity()['Account']}/{QUEUE_NAME}"

# Command pattern
# https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions


class EntryCommand(BaseModel):
    type: Literal["entry"] = "entry"
    prompt: str


class UnknownCommand(BaseModel):
    type: Literal["unknown"] = "unknown"


Command = Union[EntryCommand, UnknownCommand]


class Request(BaseModel):
    command: Command = Field(..., discriminator="type")
    id: str


# Utility functions


def send_command(command: Command):
    request = Request(id=current_session(), command=command)
    return sqs.send_message(QueueUrl=queue_url, MessageBody=request.model_dump_json())


def send_batch(commands: list[Command]):
    return sqs.send_message_batch(
        QueueUrl=queue_url,
        Entries=[
            {
                "Id": str(i),
                "MessageBody": Request(
                    id=current_session(), command=command
                ).model_dump_json(),
            }
            for i, command in enumerate(commands)
        ],
    )
