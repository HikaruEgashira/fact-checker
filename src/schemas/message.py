import os
from typing import Literal
from aws_lambda_powertools.utilities.parser import BaseModel
from schemas.aws import session

QUEUE_NAME = os.environ.get("QUEUE_NAME") or "fact-checker-queue"
REGION = os.environ.get("AWS_REGION") or "ap-northeast-1"

sqs = session.client("sqs")
sts = session.client("sts")

queue_url = f"https://sqs.{REGION}.amazonaws.com/{sts.get_caller_identity()['Account']}/{QUEUE_NAME}"


ActionType = Literal["execute"]


class ExecuteMessage(BaseModel):
    type: ActionType = "execute"
    text: str
    task_id: str


def send_message(message: BaseModel):
    message_raw = message.model_dump_json()
    return sqs.send_message(QueueUrl=queue_url, MessageBody=message_raw)
