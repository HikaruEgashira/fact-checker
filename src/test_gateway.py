from time import sleep
import urllib3
import json
import os

from schemas.state import State

API_ENDPOINT = os.environ.get("API_ENDPOINT")
if not API_ENDPOINT:
    raise ValueError("API_ENDPOINT environment variable is required")


def test_gateway(snapshot):
    http = urllib3.PoolManager()

    # POST /fact-check
    data = {"prompt": "The earth is round."}
    response = http.request(
        "POST",
        f"{API_ENDPOINT}/fact-check",
        body=json.dumps(data),
        headers={"Content-Type": "application/json"},
    )
    assert response.status == 200

    response_json = json.loads(response.data.decode("utf-8"))
    state = State(**response_json)

    assert state.status == "pending"
    state.id = "***"
    assert state.model_dump_json() == snapshot(name="POST /fact-check")

    # GET /fact-check/{id}
    state_id = json.loads(response.data)["id"]
    retry = 10
    state: State | None = None
    while retry > 0:
        response = http.request("GET", f"{API_ENDPOINT}/fact-check/{state_id}")
        response_json = json.loads(response.data.decode("utf-8"))
        state = State(**response_json)
        if state.status == "completed":
            break

        print("Waiting for state to complete...")
        sleep(3)
        retry -= 1
    assert response.status == 200

    assert state is not None
    assert state.status == "completed"
    state.id = "***"
    assert state.model_dump_json() == snapshot(name="GET /fact-check/{id}")
