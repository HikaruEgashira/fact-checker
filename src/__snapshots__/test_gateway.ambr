# serializer version: 1
# name: test_gateway[GET /fact-check/{id}]
  '{"id":"***","status":"completed","output":"accurate"}'
# ---
# name: test_gateway[POST /fact-check]
  '{"id":"***","status":"pending","output":""}'
# ---
